### PR TITLE
Add SSH manager extension

### PR DESCRIPTION
**Pull request by scopsy**
## Description
A new extension to manage ssh connection and easily create a new ssh session.

## Type of change
- New extension

## Screencast
[Loom](https://www.loom.com/share/01f8ce1f0b824806a7e6b6edb04daa44)

## Checklist
- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)

### CODE REVIEW

File: extensions/ssh-manager/package.json [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  "name": "ssh-manager",
+  "title": "SSH Connection Manager",
+  "description": "Quickly connect and manage multiple SSH connections",
+  "icon": "command-icon.png",
+  "author": "dimagrossman",
+  "license": "MIT",
+  "commands": [
+    {
+      "name": "create",
+      "title": "Create New SSH Connection",
```
Comment (pernielsentikaer) [inline extensions/ssh-manager/package.json:12]:
```suggestion
      "title": "Create New Connection",
      "subtitle": "SSH Connection Manager",
```

File: extensions/ssh-manager/package.json [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      "description": "Add a new connection to easily access it later",
+      "mode": "view"
+    },
+    {
+      "name": "show",
+      "title": "Show SSH Connections",
```
Comment (pernielsentikaer) [inline extensions/ssh-manager/package.json:18]:
```suggestion
      "title": "Show Connections",
      "subtitle": "SSH Connection Manager",
```

File: extensions/ssh-manager/package.json [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      "description": "Quickly access pre-defined ssh connections and connect to them easily",
+      "mode": "view"
+    }
+  ],
+  "dependencies": {
+    "@raycast/api": "^1.27.0",
```
Comment (pernielsentikaer) [inline extensions/ssh-manager/package.json:24]:
```suggestion
    "@raycast/api": "^1.27.1",
```

File: extensions/ssh-manager/src/create.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import { ActionPanel, Form, showHUD, SubmitFormAction, useNavigation } from "@raycast/api";
```
Comment (pernielsentikaer) [inline extensions/ssh-manager/src/create.tsx:1]:
```suggestion
import { ActionPanel, Form, showHUD, SubmitFormAction, useNavigation, randomId } from "@raycast/api";
```
You could use randomId from Raycast API instead

File: extensions/ssh-manager/src/create.tsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import { ISSHConnection } from "./types";
+import { getConnections, saveConnections } from "./storage.api";
+import { createId } from "./utils";
```
Comment (pernielsentikaer) [inline extensions/ssh-manager/src/create.tsx:4]:
```suggestion
import { createId } from "./utils";
```
Could be unused if you change

File: extensions/ssh-manager/src/create.tsx [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+export default function Main() {
+  const { pop } = useNavigation();
+
+  async function saveConnection(connection: ISSHConnection) {
+    const existingConnections = await getConnections();
+    existingConnections.push({ ...connection, id: createId() });
```
Comment (pernielsentikaer) [inline extensions/ssh-manager/src/create.tsx:11]:
```suggestion
    existingConnections.push({ ...connection, id: randomId() });
```

File: extensions/ssh-manager/src/create.tsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+  return (
+    <Form
+      actions={
+        <ActionPanel>
+          <SubmitFormAction title="Save" onSubmit={(values) => saveConnection(values as ISSHConnection)} />
+        </ActionPanel>
+      }
+    >
+      <Form.TextField id="name" title="Connection name" />
```
Comment (pernielsentikaer) [inline extensions/ssh-manager/src/create.tsx:27]:
```suggestion
      <Form.TextField id="name" title="Connection Name" />
```

File: extensions/ssh-manager/src/create.tsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+      <Form.TextField id="address" title="Server Address" placeholder={"A resolvable DNS name or IP"} />
+      <Form.TextField id="user" title="Username" placeholder={"A username to authenticate with"} />
+      <Form.TextField
+        id="key"
+        title="SSH Key location (optional)"
```
Comment (pernielsentikaer) [inline extensions/ssh-manager/src/create.tsx:32]:
```suggestion
        title="SSH Key Location (optional)"
```

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @scopsy 👋 

Thanks for your contribution 💯 💪 

Besides the suggestions, I have some feedback:
- In the newest version of Raycast API (^1.27.1) it's possible to make an “Open with preference” – could that work somehow to make it work with other terminals? 
- Your extension have [empty state flicker](https://developers.raycast.com/basics/prepare-an-extension-for-store#empty-states) – could you look into this?

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by scopsy** [comment]
Thanks @pernielsentikaer pushed the changes.
I'm not sure about what do you mean when saying Open with preference, couldn't find it anywhere in the docs

**Comment by pernielsentikaer** [comment]
> Thanks @pernielsentikaer pushed the changes. I'm not sure about what do you mean when saying Open with preference, couldn't find it anywhere in the docs

Check out the examples, [right around here](https://github.com/raycast/extensions/blob/main/examples/api-examples/package.json#L162) - maybe it can be used to set default terminal somehow 😄

**Comment by fe9lix** [comment]
Hey @scopsy, what's the status of this PR? It's been stale for a while...

**Comment by scopsy** [comment]
Hi, @fe9lix on my end I think I've done all the necessary changes needed. Apart from the open with preference part which I'm not sure how to implement and why it's needed. The current version works great locally.

**Comment by fe9lix** [comment]
@pernielsentikaer Could you do a re-review on this?

File: extensions/ssh-manager/src/create.tsx [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        <ActionPanel>
+          <SubmitFormAction title="Save" onSubmit={(values) => saveConnection(values as ISSHConnection)} />
+        </ActionPanel>
+      }
+    >
+      <Form.TextField id="name" title="Connection Name" />
+      <Form.TextField id="address" title="Server Address" placeholder={"A resolvable DNS name or IP"} />
+      <Form.TextField id="user" title="Username" placeholder={"A username to authenticate with"} />
+      <Form.TextField
+        id="key"
```
Comment (pernielsentikaer) [inline extensions/ssh-manager/src/create.tsx:30]:
```suggestion
        id="sshKey"
```
Looks like it has the wrong name and won't work without this

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @scopsy 👋 

I asked because it's opening in “terminal” which is not my default terminal, so it would be cool if other terminal would work, of course, it's not required at all 😄 

I did a new review and found a thing not working, could you address this?

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by sxn** [comment]
@pernielsentikaer @scopsy I've fixed the error you pointed out & updated the extension to the latest API version. Mind re-reviewing? 😸

**Comment by pernielsentikaer** [comment]
Awesome, works as expected 🔥

**Pull request merged**